File: Sheets/src/neetcode/practice/Trees/_3_diameterOfBinaryTree.java
package neetcode.practice.Trees;

class Solution {
  int res = 0;

  public int dfs(TreeNode root, Procedure proc) {
    if (root == null)
      return 0;

    int leftHeight = dfs(root.left, proc);
    int rightHeight = dfs(root.right, proc);

    proc.run(leftHeight + rightHeight);

    return 1 + Math.max(leftHeight, rightHeight);
  }

  public int diameterOfBinaryTree(TreeNode root) {
    Procedure proc = update -> this.res = Math.max(this.res, update);
    this.dfs(root, proc);
    return this.res;
  }
}

interface Procedure {
  public void run(int update);
}

File: Sheets/src/neetcode/practice/Trees/_3_DiameterOfBinaryTree.cc
#include "TreeNode.hh"
#include <algorithm>

class Solution
{
private:
  int diameter_{};

  int dfs_(TreeNode *root)
  {
    if (root == nullptr)
      return 0;

    int left_incl_height  = dfs_(root->left);
    int right_incl_height = dfs_(root->right);

    diameter_ = std::max(diameter_, left_incl_height + right_incl_height);

    return 1 + std::max(left_incl_height, right_incl_height);
  }

public:
  int diameterOfBinaryTree(TreeNode *root)
  {
    diameter_ = 0;
    dfs_(root);
    return diameter_;
  }
};
